sw/device/tests/sram_ctrl_scrambled_access_test.c

```diff
@@ -0,0 +1,205 @@
+  return data_region_read32(region2, 0) == expected_pattern;
+}
+
+/**
+ * Reads a pattern from RAM, and checks for inequality.
+ *
+ * The values from the start and the end of RAM are checked.
+ */
+static bool read_pattern_from_ram_neq(const sram_ctrl_t *sram,
+                                      uint32_t expected_pattern) {
+  LOG_INFO("Reading [%x] from the start of RAM = %x", expected_pattern,
+           sram->ram_start);
+  data_region_t region1 = data_region_from_addr(sram->ram_start);
+  if (data_region_read32(region1, 0) == expected_pattern) {
+    return false;
+  }
+
+  LOG_INFO("Reading [%x] from the end of RAM = %x", expected_pattern,
+           sram->ram_end);
+  data_region_t region2 = data_region_from_addr(sram->ram_end);
+  return data_region_read32(region2, 0) != expected_pattern;
+}
+
+/**
+ * Tests that the written data to RAM can be read.
+ *
+ * This is a pre-requisite test, RAM can be written and read successfully.
+ * The test write-read check one pattern, and then the second, which prevents
+ * a false-negative in case of the value of the memory address under test is
+ * already initialised to one of the patterns.
+ */
+static void test_ram_write_read_pattern(const sram_ctrl_t *sram) {
+  // Write first pattern to the start and the end of RAM.
+  write_pattern_to_ram(sram, 0xA5A5A5A5);
+  CHECK(read_pattern_from_ram_eq(sram, 0xA5A5A5A5),
+        "Read-Write first pattern failed");
+
+  // Write second pattern to the start and the end of RAM.
+  write_pattern_to_ram(sram, 0x5A5A5A5A);
+  CHECK(read_pattern_from_ram_eq(sram, 0x5A5A5A5A),
+        "Read-Write second pattern failed");
+}
+
+/**
+ * Tests RAM Scrambling.
+ *
+ * This test first writes a pattern to the start and the end of RAM. It then
+ * requests a new scrambling key via dif_sram_ctrl (which scrambles RAM
+ * addresses as well as the data). Finally, it reads back the RAM addresses
+ * an checks them against the original pattern. The test is successful if
+ * pattern doesn't match.
+ */
+bool test_ram_scrambling(const sram_ctrl_t *sram) {
+  // Write the pattern at the beginning and the end of RAM.
+  write_pattern_to_ram(sram, 0xA5A5A5A5);
+
+  dif_result_t result = dif_sram_ctrl_request_new_key(&sram->dif);
+  CHECK(result == kDifOk, "Scramble error: %d", result);
+
+  // Loop until new scrambling key has been obtained.
+  LOG_INFO("Scrambling Retention RAM!");
+  dif_sram_ctrl_status_bitfield_t status;
+  do {
+    dif_result_t result = dif_sram_ctrl_get_status(&sram->dif, &status);
+    CHECK(result == kDifOk, "Error obtaining status: %d", result);
+  } while (!(status & kDifSramCtrlStatusScrKeyValid));
+
+  // Read the first pattern at the beginning and the end of RAM.
+  // It should not match the originally written pattern.
+  return read_pattern_from_ram_neq(sram, 0xA5A5A5A5);
+}
+
+bool test_main(void) {
+  (void)sram_main;
+  sram_ctrl_init(&sram_ret);
+
+  LOG_INFO("Executing the Retention RAM Read-Write test!");
```

Comment:
tiny nit: maybe drop the exclamation marks from messages printed during normal execution of the test.

Reply:
Otherwise it looks that we are too happy that the tests are executing :)